0.3.1 has a breaking change.

Hi there,

One of my CI jobs failed today with the following compilation error:
```
error[E0107]: this struct takes 2 generic arguments but 1 generic argument was supplied
   --> rest/src/main.rs:116:24
    |
116 | fn ratelimit_post() -> GovernorConfig<PeerIpKeyExtractor> {
    |                        ^^^^^^^^^^^^^^ ------------------ supplied 1 generic argument
    |                        |
    |                        expected 2 generic arguments
    |
note: struct defined here, with 2 generic parameters: `K`, `M`
   --> /root/.cargo/registry/src/index.crates.io-e139d0d48fed7772/actix-governor-0.3.1/src/lib.rs:388:12
    |
388 | pub struct GovernorConfig<K: KeyExtractor, M: RateLimitingMiddleware<QuantaInstant>> {
    |            ^^^^^^^^^^^^^^ -                -
help: add missing generic argument
    |
116 | fn ratelimit_post() -> GovernorConfig<PeerIpKeyExtractor, M> {
    |                                                         +++

```
This compiled fine on `0.3.0`, but not on `0,3.1`. Code written for `0.3.0` should be compatible with `0.3.1` according to semver (I think).

IMO `0.3.1` should be yanked and republished as `0.4.0`.

Thanks!

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "actix-governor"
-version = "0.3.1"
+version = "0.3.2"
 authors = ["Aaron Erhardt <aaron.erhardt@t-online.de>"]
 edition = "2021"
 description = "A rate-limiting middleware for actix-web backed by the governor crate"

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -134,14 +134,9 @@
 #[cfg(test)]
 mod tests;
 
-use governor::{
-    clock::{DefaultClock, QuantaInstant},
-    middleware::{NoOpMiddleware, RateLimitingMiddleware, StateInformationMiddleware},
-    state::keyed::DefaultKeyedStateStore,
-    Quota, RateLimiter,
-};
+use governor::{clock::DefaultClock, state::keyed::DefaultKeyedStateStore, Quota, RateLimiter};
 
-use std::{cell::RefCell, marker::PhantomData, num::NonZeroU32, rc::Rc, sync::Arc, time::Duration};
+use std::{cell::RefCell, num::NonZeroU32, rc::Rc, sync::Arc, time::Duration};
 
 use actix_web::dev::{Service, ServiceRequest, ServiceResponse, Transform};
 use actix_web::http::Method;
@@ -151,8 +146,7 @@ use futures::future;
 mod key_extractor;
 mod service;
 
-type SharedRateLimiter<Key, M> =
-    Arc<RateLimiter<Key, DefaultKeyedStateStore<Key>, DefaultClock, M>>;
+type SharedRateLimiter<Key> = Arc<RateLimiter<Key, DefaultKeyedStateStore<Key>, DefaultClock>>;
 
 pub use key_extractor::{GlobalKeyExtractor, KeyExtractor, PeerIpKeyExtractor};
 
@@ -189,31 +183,25 @@ const DEFAULT_BURST_SIZE: u32 = 8;
 ///     .unwrap();
 /// ```
 #[derive(Debug, Eq)]
-pub struct GovernorConfigBuilder<K: KeyExtractor, M: RateLimitingMiddleware<QuantaInstant>> {
+pub struct GovernorConfigBuilder<K: KeyExtractor> {
     period: Duration,
     burst_size: u32,
     methods: Option<Vec<Method>>,
     key_extractor: K,
-    middleware: PhantomData<M>,
 }
 
-impl<K: KeyExtractor, M: RateLimitingMiddleware<QuantaInstant>> Clone
-    for GovernorConfigBuilder<K, M>
-{
+impl<K: KeyExtractor> Clone for GovernorConfigBuilder<K> {
     fn clone(&self) -> Self {
         Self {
             period: self.period,
             burst_size: self.burst_size,
             methods: self.methods.clone(),
             key_extractor: self.key_extractor.clone(),
-            middleware: self.middleware,
         }
     }
 }
 
-impl<K: KeyExtractor + PartialEq, M: RateLimitingMiddleware<QuantaInstant>> PartialEq
-    for GovernorConfigBuilder<K, M>
-{
+impl<K: KeyExtractor + PartialEq> PartialEq for GovernorConfigBuilder<K> {
     fn eq(&self, other: &Self) -> bool {
         self.period == other.period
             && self.burst_size == other.burst_size
@@ -222,7 +210,7 @@ impl<K: KeyExtractor + PartialEq, M: RateLimitingMiddleware<QuantaInstant>> Part
     }
 }
 
-impl Default for GovernorConfigBuilder<PeerIpKeyExtractor, NoOpMiddleware> {
+impl Default for GovernorConfigBuilder<PeerIpKeyExtractor> {
     /// The default configuration which is suitable for most services.
     /// Allows burst with up to eight requests and replenishes one element after 500ms, based on peer IP.
     /// The values can be modified by calling other methods on this struct.
@@ -231,14 +219,13 @@ impl Default for GovernorConfigBuilder<PeerIpKeyExtractor, NoOpMiddleware> {
     }
 }
 
-impl<M: RateLimitingMiddleware<QuantaInstant>> GovernorConfigBuilder<PeerIpKeyExtractor, M> {
+impl GovernorConfigBuilder<PeerIpKeyExtractor> {
     pub fn const_default() -> Self {
         GovernorConfigBuilder {
             period: DEFAULT_PERIOD,
             burst_size: DEFAULT_BURST_SIZE,
             methods: None,
             key_extractor: PeerIpKeyExtractor,
-            middleware: PhantomData,
         }
     }
     /// Set the interval after which one element of the quota is replenished.
@@ -280,7 +267,7 @@ impl<M: RateLimitingMiddleware<QuantaInstant>> GovernorConfigBuilder<PeerIpKeyEx
     }
 }
 
-impl<K: KeyExtractor, M: RateLimitingMiddleware<QuantaInstant>> GovernorConfigBuilder<K, M> {
+impl<K: KeyExtractor> GovernorConfigBuilder<K> {
     /// Set the interval after which one element of the quota is replenished.
     ///
     /// **The interval must not be zero.**
@@ -331,13 +318,12 @@ impl<K: KeyExtractor, M: RateLimitingMiddleware<QuantaInstant>> GovernorConfigBu
     pub fn key_extractor<K2: KeyExtractor>(
         &mut self,
         key_extractor: K2,
-    ) -> GovernorConfigBuilder<K2, M> {
+    ) -> GovernorConfigBuilder<K2> {
         GovernorConfigBuilder {
             period: self.period,
             burst_size: self.burst_size,
             methods: self.methods.to_owned(),
             key_extractor,
-            middleware: PhantomData,
         }
     }
 
@@ -351,30 +337,26 @@ impl<K: KeyExtractor, M: RateLimitingMiddleware<QuantaInstant>> GovernorConfigBu
     ///
     /// [`methods`]: crate::GovernorConfigBuilder::methods()
     /// [`use_headers`]: Self::use_headers
-    pub fn use_headers(&mut self) -> GovernorConfigBuilder<K, StateInformationMiddleware> {
+    pub fn use_headers(&mut self) -> GovernorConfigBuilder<K> {
         GovernorConfigBuilder {
             period: self.period,
             burst_size: self.burst_size,
             methods: self.methods.to_owned(),
             key_extractor: self.key_extractor.clone(),
-            middleware: PhantomData,
         }
     }
 
     /// Finish building the configuration and return the configuration for the middleware.
     /// Returns `None` if either burst size or period interval are zero.
-    pub fn finish(&mut self) -> Option<GovernorConfig<K, M>> {
+    pub fn finish(&mut self) -> Option<GovernorConfig<K>> {
         if self.burst_size != 0 && self.period.as_nanos() != 0 {
             Some(GovernorConfig {
                 key_extractor: self.key_extractor.clone(),
-                limiter: Arc::new(
-                    RateLimiter::keyed(
-                        Quota::with_period(self.period)
-                            .unwrap()
-                            .allow_burst(NonZeroU32::new(self.burst_size).unwrap()),
-                    )
-                    .with_middleware::<M>(),
-                ),
+                limiter: Arc::new(RateLimiter::keyed(
+                    Quota::with_period(self.period)
+                        .unwrap()
+                        .allow_burst(NonZeroU32::new(self.burst_size).unwrap()),
+                )),
                 methods: self.methods.clone(),
             })
         } else {
@@ -385,13 +367,13 @@ impl<K: KeyExtractor, M: RateLimitingMiddleware<QuantaInstant>> GovernorConfigBu
 
 #[derive(Debug)]
 /// Configuration for the Governor middleware.
-pub struct GovernorConfig<K: KeyExtractor, M: RateLimitingMiddleware<QuantaInstant>> {
+pub struct GovernorConfig<K: KeyExtractor> {
     key_extractor: K,
-    limiter: SharedRateLimiter<K::Key, M>,
+    limiter: SharedRateLimiter<K::Key>,
     methods: Option<Vec<Method>>,
 }
 
-impl<K: KeyExtractor, M: RateLimitingMiddleware<QuantaInstant>> Clone for GovernorConfig<K, M> {
+impl<K: KeyExtractor> Clone for GovernorConfig<K> {
     fn clone(&self) -> Self {
         GovernorConfig {
             key_extractor: self.key_extractor.clone(),
@@ -401,7 +383,7 @@ impl<K: KeyExtractor, M: RateLimitingMiddleware<QuantaInstant>> Clone for Govern
     }
 }
 
-impl Default for GovernorConfig<PeerIpKeyExtractor, NoOpMiddleware> {
+impl Default for GovernorConfig<PeerIpKeyExtractor> {
     /// The default configuration which is suitable for most services.
     /// Allows bursts with up to eight requests and replenishes one element after 500ms, based on peer IP.
     fn default() -> Self {
@@ -409,7 +391,7 @@ impl Default for GovernorConfig<PeerIpKeyExtractor, NoOpMiddleware> {
     }
 }
 
-impl<M: RateLimitingMiddleware<QuantaInstant>> GovernorConfig<PeerIpKeyExtractor, M> {
+impl GovernorConfig<PeerIpKeyExtractor> {
     /// A default configuration for security related services.
     /// Allows bursts with up to two requests and replenishes one element after four seconds, based on peer IP.
     ///
@@ -421,7 +403,6 @@ impl<M: RateLimitingMiddleware<QuantaInstant>> GovernorConfig<PeerIpKeyExtractor
             burst_size: 2,
             methods: None,
             key_extractor: PeerIpKeyExtractor,
-            middleware: PhantomData,
         }
         .finish()
         .unwrap()
@@ -429,15 +410,15 @@ impl<M: RateLimitingMiddleware<QuantaInstant>> GovernorConfig<PeerIpKeyExtractor
 }
 
 /// Governor middleware factory.
-pub struct Governor<K: KeyExtractor, M: RateLimitingMiddleware<QuantaInstant>> {
+pub struct Governor<K: KeyExtractor> {
     key_extractor: K,
-    limiter: SharedRateLimiter<K::Key, M>,
+    limiter: SharedRateLimiter<K::Key>,
     methods: Option<Vec<Method>>,
 }
 
-impl<K: KeyExtractor, M: RateLimitingMiddleware<QuantaInstant>> Governor<K, M> {
+impl<K: KeyExtractor> Governor<K> {
     /// Create new governor middleware factory from configuration.
-    pub fn new(config: &GovernorConfig<K, M>) -> Self {
+    pub fn new(config: &GovernorConfig<K>) -> Self {
         Governor {
             key_extractor: config.key_extractor.clone(),
             limiter: config.limiter.clone(),
@@ -446,7 +427,7 @@ impl<K: KeyExtractor, M: RateLimitingMiddleware<QuantaInstant>> Governor<K, M> {
     }
 }
 
-impl<S, B, K> Transform<S, ServiceRequest> for Governor<K, NoOpMiddleware>
+impl<S, B, K> Transform<S, ServiceRequest> for Governor<K>
 where
     K: KeyExtractor,
     S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
@@ -454,12 +435,12 @@ where
 {
     type Response = ServiceResponse<B>;
     type Error = Error;
-    type Transform = GovernorMiddleware<S, K, NoOpMiddleware>;
+    type Transform = GovernorMiddleware<S, K>;
     type InitError = ();
     type Future = future::Ready<Result<Self::Transform, Self::InitError>>;
 
     fn new_transform(&self, service: S) -> Self::Future {
-        future::ok(GovernorMiddleware::<S, K, NoOpMiddleware> {
+        future::ok(GovernorMiddleware::<S, K> {
             service: Rc::new(RefCell::new(service)),
             key_extractor: self.key_extractor.clone(),
             limiter: self.limiter.clone(),
@@ -468,32 +449,9 @@ where
     }
 }
 
-impl<S, B, K> Transform<S, ServiceRequest> for Governor<K, StateInformationMiddleware>
-where
-    K: KeyExtractor,
-    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
-    B: MessageBody,
-    <S as Service<ServiceRequest>>::Future: Unpin,
-{
-    type Response = ServiceResponse<B>;
-    type Error = Error;
-    type Transform = GovernorMiddleware<S, K, StateInformationMiddleware>;
-    type InitError = ();
-    type Future = future::Ready<Result<Self::Transform, Self::InitError>>;
-
-    fn new_transform(&self, service: S) -> Self::Future {
-        future::ok(GovernorMiddleware::<S, K, StateInformationMiddleware> {
-            service: Rc::new(RefCell::new(service)),
-            key_extractor: self.key_extractor.clone(),
-            limiter: self.limiter.clone(),
-            methods: self.methods.clone(),
-        })
-    }
-}
-
-pub struct GovernorMiddleware<S, K: KeyExtractor, M: RateLimitingMiddleware<QuantaInstant>> {
+pub struct GovernorMiddleware<S, K: KeyExtractor> {
     service: std::rc::Rc<std::cell::RefCell<S>>,
     key_extractor: K,
-    limiter: SharedRateLimiter<K::Key, M>,
+    limiter: SharedRateLimiter<K::Key>,
     methods: Option<Vec<Method>>,
 }

--- a/src/service.rs
+++ b/src/service.rs
@@ -3,7 +3,6 @@ use actix_web::http::header::{HeaderName, HeaderValue};
 use actix_web::{body::MessageBody, error, Error};
 use futures::future;
 use governor::clock::{Clock, DefaultClock};
-use governor::middleware::{NoOpMiddleware, StateInformationMiddleware};
 
 use std::future::Future;
 use std::marker::Unpin;
@@ -12,7 +11,7 @@ use std::task::{Context, Poll};
 
 use crate::{GovernorMiddleware, KeyExtractor};
 
-impl<S, B, K> Service<ServiceRequest> for GovernorMiddleware<S, K, NoOpMiddleware>
+impl<S, B, K> Service<ServiceRequest> for GovernorMiddleware<S, K>
 where
     K: KeyExtractor,
     S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
@@ -146,87 +145,6 @@ where
                 }
                 Err(err) => Err(err),
             }),
-        }
-    }
-}
-
-/// Implementation using rate limit headers
-impl<S, B, K> Service<ServiceRequest> for GovernorMiddleware<S, K, StateInformationMiddleware>
-where
-    K: KeyExtractor,
-    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
-    B: MessageBody,
-    S::Future: Unpin,
-{
-    type Response = S::Response;
-    type Error = S::Error;
-    type Future = future::Either<
-        future::Ready<Result<ServiceResponse<B>, actix_web::Error>>,
-        future::Either<RateLimitHeaderFut<S::Future>, WhitelistedHeaderFut<S::Future>>,
-    >;
-
-    fn poll_ready(&self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        self.service.poll_ready(cx)
-    }
-
-    fn call(&self, req: ServiceRequest) -> Self::Future {
-        if let Some(configured_methods) = &self.methods {
-            if !configured_methods.contains(req.method()) {
-                // The request method is not configured, we're ignoring this one.
-                let fut = self.service.call(req);
-                return future::Either::Right(future::Either::Right(WhitelistedHeaderFut {
-                    future: fut,
-                }));
-            }
-        }
-
-        // Use the provided key extractor to extract the rate limiting key from the request.
-        match self.key_extractor.extract(&req) {
-            // Extraction worked, let's check if rate limiting is needed.
-            Ok(key) => match self.limiter.check_key(&key) {
-                Ok(snapshot) => {
-                    let fut = self.service.call(req);
-                    future::Either::Right(future::Either::Left(RateLimitHeaderFut {
-                        future: fut,
-                        burst_size: snapshot.quota().burst_size().get(),
-                        remaining_burst_capacity: snapshot.remaining_burst_capacity(),
-                    }))
-                }
-
-                Err(negative) => {
-                    let wait_time = negative
-                        .wait_time_from(DefaultClock::default().now())
-                        .as_secs();
-
-                    #[cfg(feature = "log")]
-                    {
-                        let key_name = match self.key_extractor.key_name(&key) {
-                            Some(n) => format!(" [{}]", &n),
-                            None => "".to_owned(),
-                        };
-                        log::info!(
-                            "Rate limit exceeded for {}{}, quota reset in {}s",
-                            self.key_extractor.name(),
-                            key_name,
-                            &wait_time
-                        );
-                    }
-
-                    let (body, content_type) = self.key_extractor.response_error_content(&negative);
-                    let response = actix_web::HttpResponse::TooManyRequests()
-                        .insert_header(content_type)
-                        .insert_header(("x-ratelimit-after", wait_time))
-                        .insert_header(("x-ratelimit-limit", negative.quota().burst_size().get()))
-                        .insert_header(("x-ratelimit-remaining", 0))
-                        .body(body.clone());
-                    future::Either::Left(future::err(
-                        error::InternalError::from_response(body, response).into(),
-                    ))
-                }
-            },
-
-            // Extraction failed, stop right now.
-            Err(e) => future::Either::Left(future::err(self.key_extractor.response_error(e))),
         }
     }
 }

